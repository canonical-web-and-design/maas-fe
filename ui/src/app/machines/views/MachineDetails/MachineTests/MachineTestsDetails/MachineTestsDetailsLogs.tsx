import { useState } from "react";

import { Code, Tabs } from "@canonical/react-components";

import type { ScriptResultData } from "app/store/scriptresult/types";
import { ScriptResultDataType } from "app/store/scriptresult/types";

type Props = {
  log: ScriptResultData;
};

const MachineTestsDetailsLogs = ({ log }: Props): JSX.Element => {
  const [activeTab, setActiveTab] = useState<string>("combined");

  // Ideally should be buttons, see https://github.com/canonical-web-and-design/vanilla-framework/issues/3532
  const links = [
    ScriptResultDataType.COMBINED,
    ScriptResultDataType.STDOUT,
    ScriptResultDataType.STDERR,
    ScriptResultDataType.RESULT,
  ].map((tab) => ({
    active: activeTab === tab,
    label: tab === "result" ? "yaml" : tab,
    onClick: (e: React.MouseEvent) => {
      e.preventDefault();
      setActiveTab(tab);
    },
  }));

  const tabText = log[activeTab] || "";

  return (
    <>
      <Tabs links={links} />
      <Code data-test="log-content" className="u-no-margin--bottom">
        {tabText === "" ? "No data" : tabText}
      </Code>
    </>
  );
};

export default MachineTestsDetailsLogs;
