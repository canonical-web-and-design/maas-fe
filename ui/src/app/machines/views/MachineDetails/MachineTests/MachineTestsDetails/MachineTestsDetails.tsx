import { useEffect, useState } from "react";

import { Col, Row, Spinner, Tooltip } from "@canonical/react-components";
import { useDispatch, useSelector } from "react-redux";
import { Link, useParams, useLocation } from "react-router-dom";

import MachineTestsDetailsLogs from "./MachineTestsDetailsLogs";

import ScriptStatus from "app/base/components/ScriptStatus";
import type { RouteParams } from "app/base/types";
import machineURLs from "app/machines/urls";
import type { RootState } from "app/store/root/types";
import { actions as scriptResultActions } from "app/store/scriptresult";
import scriptResultSelectors from "app/store/scriptresult/selectors";
import type { ScriptResultResult } from "app/store/scriptresult/types";
import { ScriptResultDataType } from "app/store/scriptresult/types";

type DetailsRouteParams = RouteParams & { scriptResultId: string };

const MachineTestsDetails = (): JSX.Element | null => {
  const [fetched, setFetched] = useState(false);
  const dispatch = useDispatch();
  const params = useParams<DetailsRouteParams>();
  const { id, scriptResultId } = params;
  const returnPath = useLocation().pathname.split("/")?.[3];
  const result = useSelector((state: RootState) =>
    scriptResultSelectors.getById(state, Number(scriptResultId))
  );
  const logs = useSelector(scriptResultSelectors.logs);
  const loading = useSelector(scriptResultSelectors.loading);
  const log = logs ? logs[parseInt(scriptResultId, 10)] : null;

  useEffect(() => {
    if (!fetched) {
      dispatch(scriptResultActions.get(Number(scriptResultId)));
      setFetched(true);
    }
  }, [dispatch, scriptResultId, fetched, setFetched]);

  useEffect(() => {
    if (!(logs && logs[Number(scriptResultId)]) && result) {
      [
        ScriptResultDataType.COMBINED,
        ScriptResultDataType.STDOUT,
        ScriptResultDataType.STDERR,
        ScriptResultDataType.RESULT,
      ].forEach((type) =>
        dispatch(scriptResultActions.getLogs(result.id, type))
      );
    }
  }, [dispatch, result, logs, scriptResultId]);

  if (loading) {
    return <Spinner />;
  } else if (!result) {
    return <h4 data-test="not-found">Script result could not be found.</h4>;
  }

  if (result) {
    const hasMetrics = result.results.length > 0;
    return (
      <>
        <Row className="u-sv2">
          <Col size="8">
            <h2 className="p-heading--four">{result.name} details</h2>
          </Col>
          <Col size="4" className="u-align--right">
            <Link to={`${machineURLs.machine.index({ id })}/${returnPath}`}>
              &lsaquo; Back to test results
            </Link>
          </Col>
        </Row>
        <Row className="u-sv2">
          <Col size="6">
            <Row>
              <Col size="2">Status</Col>
              <Col size="4">
                <ScriptStatus status={result.status}>
                  {result.status_name}
                </ScriptStatus>
              </Col>
            </Row>
            <Row>
              <Col size="2">Exit status</Col>
              <Col size="4">{result.exit_status}</Col>
            </Row>
            <Row>
              <Col size="2">Tags</Col>
              <Col size="4">{result.tags}</Col>
            </Row>
          </Col>
          <Col size="6">
            <Row>
              <Col size="2">Start time</Col>
              <Col size="4">{result.started}</Col>
            </Row>
            <Row>
              <Col size="2">End time</Col>
              <Col size="4">{result.ended}</Col>
            </Row>
            <Row>
              <Col size="2">Runtime</Col>
              <Col size="4">{result.runtime}</Col>
            </Row>
          </Col>
        </Row>
        {hasMetrics ? (
          <Row>
            <Col size="12">
              <h4>Metrics</h4>
              <table role="grid" data-test="script-details-metrics">
                <tbody>
                  {result.results.map((item: ScriptResultResult) => (
                    <tr role="row" key={`metric-${item.name}`}>
                      <td role="gridcell">
                        <Tooltip message={item.description}>
                          {item.title}
                        </Tooltip>
                      </td>
                      <td role="gridcell">{item.value}</td>
                    </tr>
                  ))}
                </tbody>
              </table>
            </Col>
          </Row>
        ) : null}
        {log ? (
          <Row>
            <Col size="12">
              <h4>Output</h4>
              <MachineTestsDetailsLogs log={log} />
            </Col>
          </Row>
        ) : null}
      </>
    );
  }
  return null;
};

export default MachineTestsDetails;
