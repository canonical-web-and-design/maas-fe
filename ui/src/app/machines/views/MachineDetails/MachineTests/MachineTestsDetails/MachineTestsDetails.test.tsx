import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter, Route } from "react-router";
import configureStore from "redux-mock-store";

import MachineTestsDetails from ".";

import type { RootState } from "app/store/root/types";
import {
  machineState as machineStateFactory,
  machineDetails as machineDetailsFactory,
  rootState as rootStateFactory,
  scriptResult as scriptResultFactory,
  scriptResultResult as scriptResultResultFactory,
  scriptResultState as scriptResultStateFactory,
} from "testing/factories";

const mockStore = configureStore();

describe("MachineTestDetails", () => {
  let state: RootState;
  beforeEach(() => {
    state = rootStateFactory({
      machine: machineStateFactory({
        loaded: true,
        items: [
          machineDetailsFactory({
            locked: false,
            permissions: ["edit"],
            system_id: "abc123",
          }),
        ],
      }),
      scriptresult: scriptResultStateFactory({
        loaded: true,
        items: [scriptResultFactory()],
      }),
    });
  });

  it("renders", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <MachineTestsDetails />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("displays script result details", () => {
    const scriptResult = scriptResultFactory({ id: 1 });
    const scriptResults = [scriptResult];
    state.nodescriptresult.items = { abc123: [1] };
    state.scriptresult.items = scriptResults;
    const store = mockStore(state);

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/machine/abc123/tests/1/details"]}>
          <Route path="/machine/:id/tests/:scriptResultId/details">
            <MachineTestsDetails />
          </Route>
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find("span[data-test='status-name']").text()).toEqual(
      scriptResult.status_name
    );
  });

  it("displays script result metrics", () => {
    const metrics = scriptResultResultFactory({
      title: "test-title",
      value: "test-value",
    });
    const scriptResults = [scriptResultFactory({ id: 1, results: [metrics] })];

    state.nodescriptresult.items = { abc123: [1] };
    state.scriptresult.items = scriptResults;

    const store = mockStore(state);

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/machine/abc123/tests/1/details"]}>
          <Route path="/machine/:id/tests/:scriptResultId/details">
            <MachineTestsDetails />
          </Route>
        </MemoryRouter>
      </Provider>
    );

    expect(
      wrapper.find("table[data-test='script-details-metrics']").exists()
    ).toEqual(true);
    expect(
      wrapper
        .find("table[data-test='script-details-metrics'] tr")
        .at(0)
        .find("td")
        .at(0)
        .text()
    ).toEqual("test-title");
    expect(
      wrapper
        .find("table[data-test='script-details-metrics'] tr")
        .at(0)
        .find("td")
        .at(1)
        .text()
    ).toEqual("test-value");
  });

  it("fetches script result logs", () => {
    const scriptResults = [scriptResultFactory({ id: 1 })];

    state.nodescriptresult.items = { abc123: [1] };
    state.scriptresult.items = scriptResults;

    const store = mockStore(state);

    mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/machine/abc123/tests/1/details"]}>
          <Route path="/machine/:id/tests/:scriptResultId/details">
            <MachineTestsDetails />
          </Route>
        </MemoryRouter>
      </Provider>
    );

    expect(store.getActions()[0].type).toEqual("scriptresult/getLogs");
    expect(store.getActions()[0].payload.params.data_type).toEqual("combined");
    expect(store.getActions()[1].payload.params.data_type).toEqual("stdout");
    expect(store.getActions()[2].payload.params.data_type).toEqual("stderr");
    expect(store.getActions()[3].payload.params.data_type).toEqual("result");
  });
});
