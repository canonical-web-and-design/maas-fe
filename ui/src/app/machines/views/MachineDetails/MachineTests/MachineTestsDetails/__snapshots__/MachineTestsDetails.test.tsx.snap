// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MachineTestDetails renders 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        Object {
          "key": "testKey",
          "pathname": "/machine/abc123",
        },
      ]
    }
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "testKey",
              "pathname": "/machine/abc123",
              "search": "",
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/machine/abc123",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <CompatRouter>
        <Router
          location={
            Object {
              "hash": "",
              "key": "testKey",
              "pathname": "/machine/abc123",
              "search": "",
            }
          }
          navigationType="POP"
          navigator={
            Object {
              "action": "POP",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "key": "testKey",
                  "pathname": "/machine/abc123",
                  "search": "",
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 0,
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "key": "testKey",
                "pathname": "/machine/abc123",
                "search": "",
              },
              "push": [Function],
              "replace": [Function],
            }
          }
        >
          <Routes>
            <MachineTestsDetails>
              <h4
                data-testid="not-found"
              >
                Script result could not be found.
              </h4>
            </MachineTestsDetails>
          </Routes>
        </Router>
      </CompatRouter>
    </Router>
  </MemoryRouter>
</Provider>
`;
